src/v0/destinations/braze/util.js

```diff
@@ -0,0 +1,181 @@
+        externalIds.push(userId);
+      } else if (anonymousId) {
+        aliasIds.push(anonymousId);
+      }
+    }
+    const externalIdsToQuery = [...new Set(externalIds)];
+    const aliasIdsToQuery = [...new Set(aliasIds)];
+
+    const identifiers = [];
+    if (externalIdsToQuery.length > 0) {
+      externalIdsToQuery.forEach((externalId) => {
+        identifiers.push({
+          external_id: externalId,
+        });
+      });
+    }
+    if (aliasIdsToQuery.length > 0) {
+      aliasIdsToQuery.forEach((aliasId) => {
+        identifiers.push({
+          alias_name: aliasId,
+          alias_label: 'rudder_id',
+        });
+      });
+    }
+    const identfierChunks = _.chunk(identifiers, 50);
+
+    const chunkedUserData = await Promise.all(
+      identfierChunks.map(async (ids) => {
+        const externalIdentifiers = ids.filter((id) => id.external_id !== undefined);
+        const aliasIdentifiers = ids.filter((id) => id.alias_name !== undefined);
+
+        const startTime = Date.now();
+        const lookUpResponse = await httpPOST(
+          `${getEndpointFromConfig(inputs[0].destination)}/users/export/ids`,
+          {
+            external_ids: externalIdentifiers.map((extId) => extId.external_id),
+            user_aliases: aliasIdentifiers,
+          },
+          {
+            headers: {
+              Authorization: `Bearer ${inputs[0].destination.Config.restApiKey}`,
+            },
+            timeout: 10 * 1000,
+          },
+        );
+        const endTime = Date.now();
+        // TODO: Remove this log
+        console.log(
+          `Time taken to fetch user store: ${endTime - startTime} ms for ${ids.length} users`,
+        );
+        const processedLookUpResponse = processAxiosResponse(lookUpResponse);
+        const { users } = processedLookUpResponse.response;
+
+        return users;
+      }),
+    );
+
+    return _.flatMap(chunkedUserData);
+  },
+
+  enrichUserStore(users, store) {
+    if (isDefinedAndNotNull(users) && Array.isArray(users)) {
+      users.forEach((user) => {
+        if (user?.external_id) {
+          store.set(user.external_id, user);
+        } else if (user?.user_aliases) {
+          user.user_aliases.forEach((alias) => {
+            if (alias.alias_label === 'rudder_id') {
+              store.set(alias.alias_name, user);
+            }
+          });
+        }
+      });
+    }
+  },
+
+  getUserDataFromStore(inputUserData, store) {
+    const { external_id, user_alias } = inputUserData;
+    let storedUserData;
+    if (external_id) {
+      storedUserData = store.get(external_id);
+    } else if (user_alias) {
+      const rudderIdAlias = user_alias.alias_name;
+      storedUserData = store.get(rudderIdAlias);
+    }
+    return storedUserData;
+  },
+
+  deduplicate(userData, store) {
+    const excludeKeys = ['external_id', 'user_alias', 'appboy_id', 'braze_id', 'custom_events'];
+    let storedUserData = this.getUserDataFromStore(userData, store);
```

Comment:
is it possible that userIds repeated in this processRouterDest but they are not stored in braze? in that case, should we do local dedupe?

Reply:
I didn't understand this, we are deduping each of the users data for each of the events in the `input` array which should automatically handle this case. On that note I think we need to handle each of the transformation sequentially (need to discuss the performance impact)